### DISCUSSION TITLE
Useful resources

### DISCUSSION THREAD

**Discussion by giannissc** [hannobraun/fornjot#556] [General]
This is an exciting project and I hope it succeeds! Here are some resource you might find useful:
(I am not an expert in any of these. I was just curious at some point and I found this to explain some of the stuff I was curious about but it might not be that useful in the end.)

Boundary representation:
Solid Modelling and CAD Systems by Ian Stroud
Applied Geometry for Computer Graphics and CAD by Duncan Marsh
Boundary Representation Modelling Techniques by Ian Stroud
[NLPCAD](https://github.com/tatfook/NPLCAD/wiki/The-Programmable-Computer-Aided-Design)
[Morpheus](https://github.com/circuithub/morpheus/wiki/Resources)
https://www.semanticscholar.org/paper/Gabriel-constrained-Parametric-Surface-Ruiz-Cadavid/bc96519380a481b1f43ff64d28f3048846b411e3

Constraint modelling:
[1]: [Interactive 2D Constraint-Based Geometric Construction System](https://github.com/GraphiteEditor/Graphite/files/6676933/41d4.content.pdf)
[2]: [Geometric Constraint Solving in Parametric Computer-Aided Design](https://github.com/GraphiteEditor/Graphite/files/6676971/021001_1.pdf)
[3]: [Constraint-based solid modeling with geometric features Zhang, J.,](https://github.com/GraphiteEditor/Graphite/files/6677016/134.pdf)
[4]: [A geometric constraint engine](https://github.com/GraphiteEditor/Graphite/files/6676979/1-s2.0-000437029290012M-main.pdf)
[5]: [Using Degrees of Freedom Analysis To Solve Geometric Constraint Systems](https://github.com/GraphiteEditor/Graphite/files/6676981/112515.112566.pdf)
[6]: [Solving Geometric Constraint Systems](https://github.com/GraphiteEditor/Graphite/files/6676982/AAAI90-106.pdf)
[7]: [A Graph-constructive Approach to Solving Systems of Geometric Constraints](https://github.com/GraphiteEditor/Graphite/files/6677010/TOG97.pdf)

EDIT: Put another link to two github repos with their own links

**Comment by hannobraun** [comment]
Thank you, @giannissc, this is very helpful!

> Boundary Representation Modelling Techniques by Ian Stroud

I'm currently reading through this one. It's already been really helpful, so I'm looking forward to checking out the other ones.

---

Now I need to figure out where to put these links, so I don't forget them by the time I need them. I guess just keeping this issue open will do for now, while I figure something better.

**Comment by giannissc** [comment]
Maybe you can keep the issue here to put all resources relevant to cad modelling or create a new repo just for resources

**Comment by giannissc** [comment]
You might also find the work of matt keeter interesting as well!
https://www.mattkeeter.com/projects/mrep/
https://www.mattkeeter.com/projects/foxtrot/

**Comment by hannobraun** [comment]
> Maybe you can keep the issue here to put all resources relevant to cad modelling or create a new repo just for resources

I like to use issues to track actionable work items. Using them to track too much other stuff makes it harder to keep an overview, which is why I'd like to find a more appropriate place.

But while I do that, one open issue to collect resources doesn't hurt, so keep them coming :smile:

**Comment by follower** [comment]
Another endorsement here for the suggestion of looking at Matt Keeter's work. :)

While my ability to understand the underlying mathematics/algorithm work is limited, :D I do see a _lot_ of potential in the UX of "(node and/or code) + direct manipulation + visual feedback" as a paradigm.

This paradigm is something Matt has explored via a project progression that includes:

 1. [Antimony](https://www.mattkeeter.com/projects/antimony/3/)
 2. [libfive Studio](https://libfive.com/studio/)

So might be interesting work to look at  when Fornjot is at a stage to be focused more on the model creation UX aspect.

**Comment by hannobraun** [comment]
Thank you for the feedback, @follower!

I've used libfive Studio, but I have yet to take a look at Antimony. I'll have to check it out!

**Comment by hannobraun** [comment]
I moved this issue to a discussion. I think that's a more appropriate place, as it doesn't track any actionable work.

That doesn't mean I don't want more resources posted. Keep it coming, folks!

**Comment by hannobraun** [comment]
@rockandsalt has posted some interesting links in https://github.com/hannobraun/Fornjot/discussions/653#discussioncomment-2985932, on numerical stability:
- https://www.sciencedirect.com/science/article/pii/S1567832604000773
- https://cs.nyu.edu/exact/papers/
- https://cs.nyu.edu/exact/intro/

**Reply by rockandsalt** [reply]
https://cs.nyu.edu/exact/doc/guaranteed.pdf

**Comment by hannobraun** [comment]
@LegNeato [has recommended](https://github.com/hannobraun/Fornjot/issues/1021#issuecomment-1250979739) looking into [Differential Dataflow](https://github.com/TimelyDataflow/differential-dataflow), which looks like it might be really useful to handle updates to Fornjot models.

**Reply by LegNeato** [reply]
You can see an example of it in use in https://github.com/dtolnay/cargo-tally.  Additionally, https://github.com/vmware/database-stream-processor is built on similar mathematical foundations (ignore the specificity of the name).

**Reply by hannobraun** [reply]
Thanks for the additional information!

**Comment by hannobraun** [comment]
I came across this university course: [CS3621 Computing with Geometry](http://www.csl.mtu.edu/cs3621/www/Home.html)

Specifically the [notes](https://pages.mtu.edu/~shene/COURSES/cs3621/NOTES/notes.html) seem pretty useful. Covers some topics I already got from the literature in a more approachable way.
